openmc/model/triso.py

```diff
@@ -608,28 +783,215 @@ def create_triso_lattice(trisos, lower_left, pitch, shape, background):
+                p1, p2 = p2, p1
+            length = p2.surface.x0 - p1.surface.x0
+            center = (p1.surface.x0 + length/2, cyl.surface.y0, cyl.surface.z0)
+
+        # Calculate the parameters for a cylinder along the y-axis
+        elif axis == 'y':
+            if p1.surface.y0 > p2.surface.y0:
+                p1, p2 = p2, p1
+            length = p2.surface.y0 - p1.surface.y0
+            center = (cyl.surface.x0, p1.surface.y0 + length/2, cyl.surface.z0)
+
+        # Calculate the parameters for a cylinder along the z-axis
+        else:
+            if p1.surface.z0 > p2.surface.z0:
+                p1, p2 = p2, p1
+            length = p2.surface.z0 - p1.surface.z0
+            center = (cyl.surface.x0, cyl.surface.y0, p1.surface.z0 + length/2)
+
+        # Make sure the half-spaces are on the correct side of the surfaces
+        if cyl.side != '-' or p1.side != '+' or p2.side != '-':
+            return None
+
+        radius = cyl.surface.r
+
+        # The region is the volume of a cylinder, so create container
+        return _Cylinder(length, radius, axis, sphere_radius, center)
+
+    def sphere():
+        # Assume the simplest case where the sphere volume is the negative
+        # half-space of a sphere
+        if not isinstance(region, openmc.Halfspace):
+            return None
+
+        if not isinstance(region.surface, openmc.Sphere):
+            return None
+
+        if region.side != '-':
+            return None
+
+        radius = region.surface.r
+        center = (region.surface.x0, region.surface.y0, region.surface.z0)
+
+        # The region is the volume of a sphere, so create container
+        return _Sphere(radius, sphere_radius, center)
+
+    def spherical_shell():
+        # Assume the simplest case where the spherical shell volume is the
+        # intersection of the half-spaces of two spheres
+        if not isinstance(region, openmc.Intersection):
+            return None
+
+        if any(not isinstance(node, openmc.Halfspace) for node in region):
+            return None
+
+        if len(region) != 2:
+            return None
+
+        if any(not isinstance(node.surface, openmc.Sphere) for node in region):
+            return None
+
+        s1, s2 = sorted(region, key=attrgetter('surface.r'))
+        radius = s2.surface.r
+        inner_radius = s1.surface.r
+        center = (s1.surface.x0, s1.surface.y0, s1.surface.z0)
+
+        if center != (s2.surface.x0, s2.surface.y0, s2.surface.z0):
+            return None
+
+        if s1.side != '+' or s2.side != '-':
+            return None
+
+        # The region is the volume of a spherical shell, so create container
+        return _SphericalShell(radius, inner_radius, sphere_radius, center)
+
+    check_type('region', region, openmc.Region)
+
+    # Check whether the region matches any of the supported container shapes
+    # and create the container if it does
+    shapes = [rectangular_prism, cylinder, sphere, spherical_shell]
```

Comment:
Let me make a refactoring suggestion here: each one of these functions could be a `@classmethod` called `from_region` on each of the subclasses of `_Container`. If a region doesn't work for a given container, it could just raise a `ValueError`. Then below, the loop could be:
```Python
for cls in _Container.__subclasses__():
    try:
        return cls.from_region(region)
    except ValueError:
        pass
```